src/tokens/TokenImport.tsx

```diff
@@ -27,57 +30,149 @@ import variables from 'src/styles/variables'
 import { PasteButton } from 'src/tokens/PasteButton'
 import { tokensByIdSelector } from 'src/tokens/selectors'
 import { getTokenId } from 'src/tokens/utils'
-import networkConfig from 'src/web3/networkConfig'
-import { isAddress } from 'viem'
+import Logger from 'src/utils/Logger'
+import { publicClient } from 'src/viem'
+import networkConfig, { networkIdToNetwork } from 'src/web3/networkConfig'
+import { walletAddressSelector } from 'src/web3/selectors'
+import { Address, formatUnits, getContract, isAddress } from 'viem'
+
+const TAG = 'tokens/TokenImport'
 
 type Props = NativeStackScreenProps<StackParamList, Screens.TokenImport>
 
+const FETCH_TOKEN_DETAILS_TIMEOUT_MS = 5_000
+
+interface TokenDetails {
+  address: string
+  symbol: string
+  decimals: number
+  name: string
+  balance: bigint
+}
+
 export default function TokenImportScreen(_: Props) {
   const { t } = useTranslation()
   const dispatch = useDispatch()
 
-  const [isCompleteAddress, setIsCompleteAddress] = useState(false)
   const [tokenAddress, setTokenAddress] = useState('')
   const [tokenSymbol, setTokenSymbol] = useState('')
   const [error, setError] = useState<string | null>(null)
   const [networkId] = useState(networkConfig.defaultNetworkId)
+
+  const walletAddress = useSelector(walletAddressSelector)
   const supportedTokens = useSelector((state) => tokensByIdSelector(state, [networkId]))
 
-  const validateAddress = (address: string) => {
-    if (!address) return
-
-    if (isAddress(address)) {
-      setIsCompleteAddress(true)
-      // TODO(RET-891): if already imported, set state as AddressState.AlreadyImported
-      const tokenId = getTokenId(networkId, address.toLowerCase())
-      if (supportedTokens[tokenId]) {
-        setError(t('tokenImport.error.alreadySupported'))
-      } else {
-        setError(null)
-      }
-    } else {
+  const validateAddress = (address: string): address is Address => {
+    if (!address) return false
+
+    if (!isAddress(address)) {
       setError(t('tokenImport.error.invalidToken'))
+      return false
+    }
+
+    // TODO(RET-891): if already imported, set state as AddressState.AlreadyImported
+    const tokenId = getTokenId(networkId, address.toLowerCase())
+    if (supportedTokens[tokenId]) {
+      setError(t('tokenImport.error.alreadySupported'))
+      ValoraAnalytics.track(AssetsEvents.import_token_already_supported, {
+        networkId,
+        tokenId,
+        tokenAddress,
+      })
+      return false
     }
-  }
 
-  const handleAddressFocus = () => {
-    setIsCompleteAddress(false)
     setError(null)
+    return true
+  }
+
+  const fetchTokenDetails = async (
+    tokenAddress: Address,
+    walletAddress: Address
```

Comment:
Ideally I would like to move it to some other place so it could be hopefully reused.

There's `fetchTokenBalancesForAddress` in `src/tokens/saga.ts` which currently fetches details via GraphQL. I was thinking of leaving `fetchTokenDetails` in that file, and just using it in `TokenImport`. What do you think?